### PR TITLE
Blockchain client updates and bugfixes

### PR DESCRIPTION
**Pull request by KPrasch**
All test pass with no skips.

- Improves time travel test utils (includes period time travel)
- Improves miner spawning test utils
- Adds Validators for stake and locktime
- Removes additional deprecated calls to .call and .transact
- Gives `ContractDeployers` agency references, and `make_agent`
- Agent and Actor API updates for `solc` and `pyevm`
- Decouple mining activities
- Locks Pipfile (+ ~1000 lines)

### CODE REVIEW

File: nkms/blockchain/eth/actors.py [hunk 1]
```diff
@@ -86,36 +86,43 @@ def locked_tokens(self,):
...
+        deposit_txhash = self.miner_agent.contract.functions.deposit(amount, periods).transact({'from': self.address})
+
         self.blockchain.wait_for_receipt(deposit_txhash)
 
         self._transactions.append((datetime.utcnow(), deposit_txhash))
 
         return deposit_txhash
 
-    def deposit(self, amount: int, locktime: int) -> Tuple[str, str]:
+    def deposit(self, amount: int, periods: int) -> Tuple[str, str]:
```
[the remark this answers is not in the document]
Comment (tuxxy) [reply nkms/blockchain/eth/actors.py]: Hmm, it makes more sense to me. But it only does as long as the duration of a period is known to the user.

`locktime_days: int` may make more sense.

File: nkms/blockchain/eth/chains.py [hunk 2]
```diff
@@ -84,41 +86,28 @@ def wait_for_receipt(self, txhash, timeout=None) -> None:
...
+        Wait the specified number of wait_hours by comparing
+        block timestamps and mines a single block.
         """
-        from nkms.blockchain.eth.actors import Miner
 
-        miners = list()
-        for address in addresses:
-            miner = Miner(miner_agent=miner_agent, address=address)
-            miners.append(miner)
+        querytime = list(filter(lambda t: bool(t), (hours, seconds, periods)))
```
Comment (michwill) [inline nkms/blockchain/eth/chains.py]: Ah, interesting way...
Can also do `if sum(map(bool, [hours, seconds, periods])) > 1:`

File: nkms/blockchain/eth/actors.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (KPrasch) [reply nkms/blockchain/eth/actors.py]: Especially during testing, the time tokens are locked is not always represented in days.  I actually do think `locktime` is better, but it lacks indication of what type or value needs to be passed.  (as `int` represented in periods)

File: nkms/blockchain/eth/chains.py (hunk 2 above)
Comment (KPrasch) [reply nkms/blockchain/eth/chains.py]: Haha, yeah, that is a bit more readable.  I just needed a quick way to easily verify that only one arg is passed in.

**Comment by KPrasch** [comment]
OK - Made all the requested changes, and rebased over master to get t he new naming scheme.

File: nucypher/blockchain/eth/actors.py [hunk 3]
```diff
@@ -3,15 +3,15 @@
 from datetime import datetime
 from typing import Tuple, List, Union
 
-from nucypher.blockchain.eth.agents import NucypherTokenAgent
+from nucypher.blockchain.eth.agents import NuCypherTokenAgent
```
[the remark this answers is not in the document]
Comment (KPrasch) [reply nucypher/blockchain/eth/actors.py]: it's an exception among an established naming convention in the codebase.

[the remark this answers is not in the document]
Comment (jMyles) [reply nucypher/blockchain/eth/actors.py]: ...but why do you want to make an exception for this class?  :-)

[the remark this answers is not in the document]
Comment (KPrasch) [reply nucypher/blockchain/eth/actors.py]: This is not an exception, there are over 100+ occurrences in 24 files of `NuCypher`.

[the remark this answers is not in the document]
Comment (jMyles) [reply nucypher/blockchain/eth/actors.py]: Ahh, I see the problem: I didn't change the pre-existing cases in my PR #268.  That's my bad.

But if you look in the description there, I proposed `Nucypher` as the class prefix (ie, the single-worded version of the package name, `nucypher`).  Doesn't that make more sense?  In python, if it were `NuCypher`, that's suggesting that it's two words and that we'd spell instances of it `nu_cypher_whatever`.

As I count, there are only 8 classes which have `NuCypher*` in their name, and that was an oversight on my part as it doesn't match the description of my work in the PR.

[the remark this answers is not in the document]
Comment (KPrasch) [reply nucypher/blockchain/eth/actors.py]: Yep - 100% agreed.

[the remark this answers is not in the document]
Comment (KPrasch) [reply nucypher/blockchain/eth/actors.py]: It was nearly 100 renames, but I think 6397fba5f1cc53cb7b66cfa240bd3465fedcacfd will do it. :-)

[the remark this answers is not in the document]
Comment (jMyles) [reply nucypher/blockchain/eth/actors.py]: Yep, looks good.  And in the great scheme of things, 100 renames isn't so bad.  :-)

[the remark this answers is not in the document]
Comment (KPrasch) [reply nucypher/blockchain/eth/actors.py]: Agreed

File: nucypher/blockchain/eth/actors.py [hunk 4]
```diff
@@ -67,7 +67,7 @@ def __init__(self, miner_agent, address):
     def __update_locked_tokens(self) -> None:
         """Query the contract for the amount of locked tokens on this miner's eth address and cache it"""
 
-        self.__locked_tokens = self.miner_agent.read().getLockedTokens(self.address)
+        self.__locked_tokens = self.miner_agent.contract.functions.getLockedTokens(self.address).call()
```
[the remark this answers is not in the document]
Comment (KPrasch) [reply nucypher/blockchain/eth/actors.py]: I have given this some thought myself.  You're right that this is the result of a changing API.  My initial impression was to wrap every contract method, but they are too numerous, and may continue to change in the future.  We once discussed an API where `.read()` would take the string name of the method, and a dictionary of the args, but I believe that is even worse than just accessing the contract object directly.  I am feeling pretty good about the `TesterChain`; but I also share your semantic concerns.  I am open to Ideas on how to best handle this situation.

[the remark this answers is not in the document]
Comment (KPrasch) [reply nucypher/blockchain/eth/actors.py]: Ideally, This syntax is only used internally, and we expose the public methods needed to operate nucypher, wrapping this logic.

**Pull request merged**